de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/ui/util/UIActiveProjectProvider.java

```diff
@@ -0,0 +1,78 @@
+package de.fu_berlin.inf.dpp.intellij.ui.util;
+
+import com.intellij.ide.DataManager;
+import com.intellij.openapi.actionSystem.DataContext;
+import com.intellij.openapi.actionSystem.DataKeys;
+import com.intellij.openapi.project.Project;
+import de.fu_berlin.inf.dpp.session.ISarosSession;
+import de.fu_berlin.inf.dpp.session.ISarosSessionManager;
+import de.fu_berlin.inf.dpp.session.ISessionLifecycleListener;
+import de.fu_berlin.inf.dpp.session.SessionEndReason;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+import org.apache.log4j.Logger;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.concurrency.Promise;
+
+/**
+ * Class to provide access to an active project object. This class should only be used for UI
+ * purposes.
+ */
+public class UIActiveProjectProvider {
+
+  private static final Logger log = Logger.getLogger(UIActiveProjectProvider.class);
+
+  private volatile ISarosSession sarosSession;
+
+  @SuppressWarnings("FieldCanBeLocal")
+  private ISessionLifecycleListener sessionLifecycleListener =
+      new ISessionLifecycleListener() {
+        @Override
+        public void sessionStarted(ISarosSession session) {
+          sarosSession = session;
+        }
+
+        @Override
+        public void sessionEnded(ISarosSession session, SessionEndReason reason) {
+          sarosSession = null;
+        }
+      };
+
+  public UIActiveProjectProvider(ISarosSessionManager sarosSessionManager) {
+    sarosSessionManager.addSessionLifecycleListener(sessionLifecycleListener);
+  }
+
+  /**
+   * Returns a current project object.
+   *
+   * <p>If there is currently a session, the session project will be returned. Otherwise, the
+   * currently focused project will be requested and returned. In this case, <code>null</code> is
+   * possible.
+   *
+   * <p><b>NOTE:</b> The project returned by this method should only be used to for UI purposes.
+   *
+   * @return a current project object or <code>null</code> if there is no session and the focused
+   *     project could not be obtained
+   */
+  @Nullable
+  public Project getProject() {
+    ISarosSession currentSession = sarosSession;
+
+    if (currentSession != null) {
+      return sarosSession.getComponent(Project.class);
+    }
+
+    Promise<DataContext> promisedDataContext =
+        DataManager.getInstance().getDataContextFromFocusAsync();
+    Promise<Project> promisedProject = promisedDataContext.then(DataKeys.PROJECT::getData);
+
+    try {
+      return promisedProject.blockingGet(1000);
```

Comment:
This should be replaced with a static final variable.

The maximum waiting period of one second (1000 milliseconds) was chosen purely by gut feeling (by guessing how long a UI action can be delayed without the user getting annoyed). This value seem to hold (I have only done minor testing) and the waiting does not seem to cause any noticeable slowdown (especially since this mechanism is only used while there is no session, where Saros does not have a lot of  UI interactions besides the Session negotiation windows).